benchmark/utils/timer_impl.hpp

```diff
@@ -195,4 +200,34 @@ class CpuTimer : public Timer {
 };
 
 
+class MpiWrappedTimer : public Timer {
+public:
+    MpiWrappedTimer(std::shared_ptr<const gko::Executor> exec,
+                    gko::experimental::mpi::communicator comm,
+                    std::shared_ptr<Timer> concrete_timer)
+        : exec_(std::move(exec)),
+          comm_(std::move(comm)),
+          concrete_timer_(std::move(concrete_timer))
+    {}
+
+protected:
+    void tic_impl() override { concrete_timer_->tic_impl(); }
+
+    double toc_impl() override
+    {
+        auto local_duration = concrete_timer_->toc_impl();
+        double duration = local_duration;
+        comm_.template all_reduce(exec_, &local_duration, &duration, 1,
+                                  MPI_MAX);
```

Comment:
Would it make sense to make the mode controllable (report maximum or local time)? Some operations we want to benchmark are bulk-synchronous anyways, so any local timer is a decent representation of global time. Will this timer also be used for the runtime breakdown in detailed mode? There this reduction will probably add a huge overhead. How about computing aggregations locally and only reducing afterwards? That would mean extending the timer interface significantly, though.

Reply:
This is to make mpi runs work with the adaptive repetitions setting. We have to use the maximum to make sure that all process actually perform the same number of repetitions. 

Also, the reduction only happens when we actually take the timing, which is not that often, or only once if we set `repetitions` to a fixed value.

I think being able to use `repetitions=auto` is very valuable, because this allows you to specify multiple cases in the input json. Otherwise, you would need to run the benchmark for each case separately. This makes processing the results a bit more annoying. (Also you need to find appropriate repetitions values for each case.)

Reply:
Ah right, I misremembered the implementation. Thanks for the clarification!